File: leetcode/src/00844_backspace_string_compare/solution.go
package leetcode

func backspaceCompare(s string, t string) bool {
	s1 := []rune(s)
	s2 := []rune(t)
	i := len(s1) - 1
	j := len(s2) - 1
	for {
		b1 := 0
		for i >= 0 && (b1 > 0 || s1[i] == '#') {
			if s1[i] == '#' {
				b1++
			} else {
				b1--
			}
			i--
		}
		b2 := 0
		for j >= 0 && (b2 > 0 || s2[j] == '#') {
			if s2[j] == '#' {
				b2++
			} else {
				b2--
			}
			j--
		}
		if i >= 0 && j >= 0 && s1[i] == s2[j] {
			i--
			j--
		} else {
			break
		}
	}
	return i == -1 && j == -1
}

File: leetcode/src/00844_backspace_string_compare/solution.rs
struct Solution;

impl Solution {
    pub fn backspace_compare(s: String, t: String) -> bool {
        Self::after_backspace(&s) == Self::after_backspace(&t)
    }

    fn after_backspace(s: &str) -> String {
        let mut st = Vec::new();
        for ch in s.chars() {
            if ch == '#' {
                st.pop();
            } else {
                st.push(ch);
            }
        }
        st.iter().collect()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() {
        let s = String::from("ab#c");
        let t = String::from("ad#c");
        assert!(Solution::backspace_compare(s, t));

        let s = String::from("ab##");
        let t = String::from("c#d#");
        assert!(Solution::backspace_compare(s, t));

        let s = String::from("a#c");
        let t = String::from("b");
        assert!(!Solution::backspace_compare(s, t));
    }
}
